rv003usb/rv003usb.h

```diff
@@ -3,6 +3,15 @@
 
 #include "usb_config.h"
 
+// Fallback and issue warning for users with the swapped pin assignments of previous versions
+#if defined(USB_DM) && defined(USB_DP)
```

Comment:
|| instead of && ?

Reply:
Benefit? If you change only one but not the other everything is doomed. DM/DP come in a pair 🤷‍♂️

Reply:
I'm not sure why we wouldn't want to warn if only one was defined

Reply:
I want to issue the warning for someone with the old setup, but also make sure it still works. For it to still work, I need both to be defined, otherwise the fallback would fail on one of the `#define`s. That's why I require both.
Sure, I could guard every fallback line with its own if, but that would be lots of code for no real use-case:
~~~c
#if defined(USB_DM) || defined(USB_DP) || defined(USB_DPU)
#warning "You usb_config.h is outdated. Please use USB_PIN_DP, USB_PIN_DM and USB_PIN_DPU instead with their respective signal (no swapping!)"
#endif
#ifdef USB_DP
#define USB_PIN_DM USB_DP
#endif
#ifdef USB_DM
#define USB_PIN_DP USB_DM
#endif
#ifdef USB_DPU
#define USB_PIN_DPU USB_DPU
#endif
~~~
If someone - for whatever reason - only changes one of the `#define`s, I prefer the code to break instead of also handling this gracefully.

Or are there any valid reasons for handling each `#define` individually?

Reply:
There's no need to check USB_DPU and I don't think anyone wants some automagic 'fix' where the existing defines are swapped.
I'm not passionate about this, but I think your code would be more correct to check if USB_DM and/or USB_DP are defined rather than only checking if both are defined. Again, not a big deal.

Reply:
> There's no need to check USB_DPU 

Actually, there is 🤔 
If someone goes with
~~~c
#define USB_PIN_DP 3
#define USB_PIN_DM 4
#define USB_DPU 5
#define USB_PORT D
~~~
it would compile without any error or warning, but the device wouldn't enumerate, as the code would look for `USB_PIN_DPU` and just disable the pull-up if it's not there.

> and I don't think anyone wants some automagic 'fix' where the existing defines are swapped

Well but that's the point of this fix?

Reply:
@BobRyan530 I changed it to OR and moved the check for DPU outside so it's independent, so the above mentioned case would not break

Reply:
I mean you don't need to check for USB_DPU being defined in this line `#if defined(USB_DM) || defined(USB_DP) || defined(USB_DPU)`
as it's already corrected here
`#ifdef USB_DPU #define USB_PIN_DPU USB_DPU #endif`

IMO this should be #error if either USP_DP or USB_DM are defined. Why bother warning if the pins just get swapped? The point of this fix is to clarify which pin connects to D+ vs D- on the USB connector and there are already examples where USB_DP and USB_DM are 'corrected' (USB_DP -> D- and visa versa). This code would compile but break those instances.

Reply:
After rethinking it, mapping USB_DM -> D+ is really just a schematic problem which can be fixed in software. Using #warning and swapping the pins does seem reasonable.

Reply:
> as it's already corrected here

previously this was inside the #if, so it was only handled if also USB_DM / USB_DP was used

> IMO this should be #error if either USP_DP or USB_DM are defined. Why bother warning if the pins just get swapped? The point of this fix is to clarify which pin connects to D+ vs D- on the USB connector

I used the #error approach in #44 and this was not well received. That's why I made it backwards compatible and just issue a #warning

> there are already examples where USB_DP and USB_DM are 'corrected' (USB_DP -> D- and visa versa). This code would compile but break those instances.

No. The code will compile and it will work. I translate USB_DP to USB_PIN_DM which is meant to be connected to D-. That's the whole point.

If you're updating RV003USB you'll get the warning but your old code should still work fine.